safehttp/form_test.go

```diff
@@ -0,0 +1,1051 @@
+			}
+			want := test.formVal
+			got := form.Int64("pizza", 0)
+			if err := form.Err(); err != nil {
+				t.Errorf(`form.Error: got "%v", want nil`, err)
+			}
+			if diff := cmp.Diff(want, got); diff != "" {
+				t.Errorf("form.Int64: got %v, want %v, diff (-want +got): \n%s", got, want, diff)
+			}
+			return Result{}
+		}, &dispatcher{})
+		recorder := httptest.NewRecorder()
+		m.HandleRequest(recorder, test.req)
+		if respStatus := recorder.Result().Status; respStatus != status200OK {
+			t.Errorf("response status: got %s, want %s", respStatus, status200OK)
+		}
+	}
+}
+
+func TestFormInvalidInt(t *testing.T) {
+	tests := []struct {
+		name    string
+		reqs    *[]*http.Request
+		err     error
+		formVal int64
+	}{
+		{
+			name: "Overflow integer in request",
+			reqs: func() *[]*http.Request {
+				getReq := httptest.NewRequest("GET", "/?pizza=9223372036854775810", nil)
+				postReq := httptest.NewRequest("POST", "/", strings.NewReader("pizza=9223372036854775810"))
+				postReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+				multipartReqBody := "--123\r\n" +
+					"Content-Disposition: form-data; name=\"pizza\"\r\n" +
+					"\r\n" +
+					"9223372036854775810\r\n" +
+					"--123--\r\n"
+				multipartReq := httptest.NewRequest("POST", "/", strings.NewReader(multipartReqBody))
+				multipartReq.Header.Set("Content-Type", `multipart/form-data; boundary="123"`)
+				res := []*http.Request{getReq, postReq, multipartReq}
+				return &res
+			}(),
+			err:     errors.New(`strconv.ParseInt: parsing "9223372036854775810": value out of range`),
+			formVal: 0,
+		},
+		{
+			name: "Not an integer in request",
+			reqs: func() *[]*http.Request {
+				getReq := httptest.NewRequest("GET", "/?pizza=diavola", nil)
+				postReq := httptest.NewRequest("POST", "/", strings.NewReader("pizza=diavola"))
+				postReq.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+				multipartReqBody := "--123\r\n" +
+					"Content-Disposition: form-data; name=\"pizza\"\r\n" +
+					"\r\n" +
+					"diavola\r\n" +
+					"--123--\r\n"
+				multipartReq := httptest.NewRequest("POST", "/", strings.NewReader(multipartReqBody))
+				multipartReq.Header.Set("Content-Type", `multipart/form-data; boundary="123"`)
+				res := []*http.Request{getReq, postReq, multipartReq}
+				return &res
+			}(),
+			err:     errors.New(`strconv.ParseInt: parsing "diavola": invalid syntax`),
+			formVal: 0,
+		},
+	}
+
+	for _, test := range tests {
+		for _, req := range *test.reqs {
+			m := NewMachinery(func(rw ResponseWriter, ir *IncomingRequest) Result {
+				form, err := getParsedForm(ir)
+				if err != nil {
+					t.Fatalf(`getParsedForm: got "%v", want nil`, err)
+				}
+				want := test.formVal
+				got := form.Int64("pizza", 0)
+				if diff := cmp.Diff(want, got); diff != "" {
+					t.Errorf("form.Int64: got %v, want %v, diff (-want +got): \n%s", got, want, diff)
+				}
+				if err := form.Err(); test.err.Error() != err.Error() {
```

Comment:
Do not compare error strings. It is fine to just check that an error is returned, unless you programatically depend on the semantics/structure of the error - which we don't.